cpp/src/rolling/rolling_detail.cuh

```diff
@@ -64,322 +66,733 @@
-      count++;
+template <typename InputType>
+struct DeviceRollingLag {
+  size_type row_offset;
+
+  // what operations do we support
+  template <typename T = InputType, aggregation::Kind O = aggregation::LAG>
+  static constexpr bool is_supported()
+  {
+    return cudf::is_fixed_width<T>();
+  }
+
+  template <typename T = InputType, typename std::enable_if_t<is_supported<T>()>* = nullptr>
+  DeviceRollingLag(size_type _row_offset) : row_offset(_row_offset)
+  {
+  }
+
+  template <typename T = InputType, typename std::enable_if_t<!is_supported<T>()>* = nullptr>
+  DeviceRollingLag(size_type _row_offset) : row_offset(_row_offset)
+  {
+    CUDF_FAIL("Invalid aggregation/type pair");
+  }
+
+  template <typename OutputType, bool has_nulls, typename T = InputType>
+  std::enable_if_t<is_supported<T>(), bool> __device__
+  operator()(column_device_view const& input,
+             column_device_view const& default_outputs,
+             mutable_column_device_view& output,
+             size_type start_index,
+             size_type end_index,
+             size_type current_index)
+  {
+    // Offsets have already been normalized.
+
+    // Check if row is invalid.
+    if (row_offset > (current_index - start_index)) {
+      // Invalid row marked. Use default value, if available.
+      if (default_outputs.size() == 0 || default_outputs.is_null(current_index)) { return false; }
+
+      output.element<OutputType>(current_index) =
+        default_outputs.element<OutputType>(current_index);
+      return true;
     }
+
+    // Not an invalid row.
+    auto index   = current_index - row_offset;
+    auto is_null = input.is_null(index);
+    if (!is_null) {
+      output.element<OutputType>(current_index) =
+        input.element<device_storage_type_t<InputType>>(index);
+    }
+    return !is_null;
   }
 
-  bool output_is_valid = (count >= min_periods);
-  // -1 will help identify null elements while gathering for Min and Max
-  // In case of count, this would be null, so doesn't matter.
-  output.element<OutputType>(current_index) = (output_is_valid) ? val_index : -1;
+  template <typename OutputType, bool has_nulls, typename T = InputType>
+  std::enable_if_t<!is_supported<T>(), bool> __device__
+  operator()(column_device_view const& input,
+             column_device_view const& default_outputs,
+             mutable_column_device_view& output,
+             size_type start_index,
+             size_type end_index,
+             size_type current_index)
+  {
+    return false;
+  }
+};
 
-  // The gather mask shouldn't contain null values, so
-  // always return zero
-  return true;
-}
+/**
+ * @brief Functor for creating a DeviceRolling operator based on input type and aggregation type.
+ */
+template <typename InputType, aggregation::Kind op>
+struct create_rolling_operator {
+  // everything else
+  auto operator()(size_type min_periods, rolling_aggregation const& agg)
+  {
+    return DeviceRolling<InputType, op>{min_periods};
+  }
+};
```

Comment:
The current code repeats the `is_supported` check for every rolling operator. It would be cleaner and less redundant to just do that once by making the general template the error case and specialize for `is_supported`.

```suggestion
template <typename InputType, aggregation::Kind op, typename Enable = void>
struct create_rolling_operator {
  auto operator()(size_type min_periods, rolling_aggregation const& agg)
  {
    CUDF_FAIL("Invalid aggregation/type pair");
  }
};

template <typename InputType, aggregation::Kind op>
struct create_rolling_operator<InputType, op, std::enable_if_t< corresponding_operator<op>::is_supported<InputType>() > > {
  auto operator()(size_type min_periods, rolling_aggregation const& agg)
  {
    return DeviceRolling<InputType, op>{min_periods};;
  }
};
```

Reply:
So the issue here is that `corresponding_operator` is used on the low level device operators (DeviceMin, DeviceMax, etc), not these mid-level rolling operators.  I could add `corresponding_rolling_operator`, although at that point the amount of code is probably bigger than what's here.

Reply:
Upon further thought, `corresponding_rolling_operator` is also the solution for the comment below https://github.com/rapidsai/cudf/pull/8158#discussion_r629556162 so it's probably the right thing to do.